File: C-Plus-Plus/cp/ZigZag.cpp
class ZigZag {
public:
    string convert(string s, int numRows) {

        if (numRows == 1) return s;

        vector<string> rows(min(numRows, int(s.size())));
        int curRow = 0;
        bool goingDown = false;

        for (char c : s) {
            rows[curRow] += c;
            if (curRow == 0 || curRow == numRows - 1) {
		goingDown = !goingDown;
	    }

            curRow += goingDown ? 1 : -1;
        }

        string ans = "";
        for (string row : rows) ans += row;
        return ans;
    }
};

/*

TIME COMPLEXITY: 0(N) where N is length of string
SPACE COMPLEXITY: o(N)
TEST CASE

INPUT
"PAYPALISHIRING"
3
OUTPUT
"PAHNAPLSIIGYIR"

*/

File: Java/cp/Zig_Zag.java
import java.util.*;
import java.lang.*;
import java.io.*;

class ZigZag {
    public String convert(String s, int numRows) {
        
        if (numRows == 1 || numRows >= s.length()) return s; 
        
        ArrayList<String> arrLL = new ArrayList<>();
        
        for (int i = 0; i < Math.min(numRows, s.length()); i++)
            arrLL.add("");
        
        int counter = 0;
        boolean b = false;
        int size = Math.min(numRows, s.length());
        for (char c : s.toCharArray()){
            String str = Character.toString(c);
            String initial = "";
           
            if (counter < arrLL.size() || arrLL.get(counter) != "") {
                initial = arrLL.get(counter);
                arrLL.add(counter, initial + str);
                arrLL.remove(counter+1);
                
            }
            if (counter == 0 || counter + 1 == numRows) b = !b;   
            counter += b ? 1 : -1;     // adding -1 in b if b is false 
                                      // adding 1 in b if b is true
        }
        
        String ans = "";
        for (int i =0; i< arrLL.size(); i++){
            ans += arrLL.get(i);
        }
        
        
        return ans;
    }
}

/*
TIME COMPLEXITY: 0(N) where n is size of string
SPACE COMPLEXITY:0(N) where n is size of string
TEST CASE

INPUT
"PAYPALISHIRING"

OUTPUT
"PAHNAPLSIIGYIR"

INPUT
"PAYPALISHIRING"
4

OUTPUT
"PINALSIGYAHRPI"

*/

